#162949069 Document API with Swagger

Teatoller opened the pull request:
#### What does this PR do?
Enables documenting of API endpoints
#### Description of Task to be completed?
- Create a branch out of develop
- Install Django rest swagger package
- Add rest_framework Swagger to INSTALLED_APPS in settings
- Update the requirments.txt file
- Add Swagger UI view in the URL path
- Customize the Swagger UI
 #### How should this be manually tested?
1. ```$ mkdir myprojects```
2. ```$ cd myprojects```
3. ```$ git clone https://github.com/andela/ah-robotics.git```
4. ```$ cd ah-robotics```

**On a Mac or Linux:**
a.  ```$ python3 -m pip install --user --upgrade pip```
b. ```$ python3 -m pip install --user virtualenv```
c. ```$ python3 -m virtualenv env```
d. ```$ source env/bin/activate```

**On Windows:**
a. ```py -m pip --version```
b. ```py -m pip install --upgrade pip```
c. ```py -m pip install --user virtualenv```
d. ```py -m virtualenv env```
e. ```.\env\Scripts\activate```

6. ```$ git checkout ft-add-swagger-docs-162949069``` 
7. ```$ pip install -r requirements.txt```

> Ensure PostgreSQL is installed and is running psql, follow these steps to create a new user and a database which will be used in the application.
 
a. Start the postgres database server using the command:
     
    pg_ctl -D /usr/local/var/postgres start
b. Run psql interactive terminal

    psql postgres
c. Create a new user:

    CREATE USER sample_username WITH PASSWORD 'sample_password';
d. Grant privileges to the user:

    ALTER USER sample_username CREATEDB;
e. Create a database:

    CREATE DATABASE sample_database_name WITH OWNER sample_username;

f. Exit psql interactive with the following command: `\q`
g. Create a .env as guided by the .env_example and add the following configuration variables

      source env/bin/activate
      export DB_NAME="database_name"
      export DB_USER="sample_username"
      export DB_PASS="sample_password"
      export DB_HOST="host_name_eg_localhost"
      export DB_PORT="port_number_eg_5432"
      export TIME_DELTA="30"
7. Head back to the terminal and source .env to set the configuration variables:
`$ source .env`
8. Ensure the database schema is up to date:
`$ python manage.py makemigrations`
8. Create the tables in the database:
`$ python manage.py migrate`
9.Run the server using the following command
`$ python3 manage.py runserver`

#### Paste URL in browser
10. ```http://127.0.0.1:8000/api/v1/docs/```

#### Any background context you want to provide?
REST API's have become more consumer-driven in this present age than it was in the past.
They have evolved to the point of influencing companies strategic decisions and how they target for business.
API's that are easy for third parties to consume becomes prime for revenue generation and business growth.
It is therefore fundamental to have customized documented API's.
#### What are the relevant pivotal tracker stories?
[#162949069](https://www.pivotaltracker.com/story/show/162949069)
#### Screenshots (if appropriate)
* Browser image displaying the Swagger UI
![screen shot 2019-01-16 at 10 29 12 am](https://user-images.githubusercontent.com/10683903/51233609-53da5480-197b-11e9-99d7-924341c7403d.png)




#### Questions:

waracci commented:
> Great job, you might want to check the format of this commit:
> [commit](https://github.com/andela/ah-robotics/pull/8/commits/3faf53065fd69030a39f26e7ef1ec27cba7ce636)
> <img alt="screenshot 2019-01-15 at 12 51 53" width="615" src="https://user-images.githubusercontent.com/6994982/51172510-6ba2d180-18c4-11e9-86b8-98647c03ebab.png">

Squash the commits for this particular commit into a single one.

jomasim commented on authors/apps/authentication/views.py:
```diff
@@ -11,6 +14,26 @@
 
 
 class RegistrationAPIView(APIView):
+    schema = AutoSchema(manual_fields=[
```
what does this class do?

johnwayodi replied:
Thanks for the question, I'll make sure to add a brief comment to explain what it does.

The `AutoSchema` class enables us to dictate the generation of the input fields, thus enabling us to make requests from the generated swagger doc, We had chosen to use it because the registration endpoint inherits from the `APIView` which can be easily linked to AutoSchema via the `schema` property of the `APIView` class. 

More info on understanding it can be found [here](https://www.django-rest-framework.org/api-guide/schemas/#autoschema)

johnwayodi replied:
@jomasim I made some changes and this is no longer used. You can review again to see the updated file structure.

johnwayodi commented:
> Great job, you might want to check the format of this commit:
> Squash the commits for this particular commit into a single one.

Thanks for flagging that, let me get to it.

Teatoller commented:
Hello Waracci,

Thank you for taking the time to review this pull request and for the valuable feedback.

Regards
Steven/Wayodi
> > Great job, you might want to check the format of this commit:
> > [commit](https://github.com/andela/ah-robotics/pull/8/commits/3faf53065fd69030a39f26e7ef1ec27cba7ce636)
> > <img alt="screenshot 2019-01-15 at 12 51 53" width="615" src="https://user-images.githubusercontent.com/6994982/51172510-6ba2d180-18c4-11e9-86b8-98647c03ebab.png">
> 
> Squash the commits for this particular commit into a single one.

Teatoller commented:
Hello, sharkdevs,

I hope you had it easy to follow through.
We have taken note and adjusted item 1 that you picked up.
With item 2 we are looking on how to autogenerate a matching API.

Regards
Steven/Wayodi

> * The body of your commit message should be separated from the header by 1 blank line. Ensure you change that.
> * I cannot run the API from the Documentation link given. This is because the example values given do not match what the API Accepts. check the screenshot below
>   ![image](https://user-images.githubusercontent.com/14871749/51174892-f5a16900-18c9-11e9-8af3-0ff50a29eab0.png)

johnwayodi commented:
> 2. I cannot run the API from the Documentation link given. This is because the example values given do not match what the API Accepts. check the screenshot below

This is an issue that we are currently working on. An update will be made once it is solved.

Teatoller commented:
@sharkdevs 
I have added additional information on the pull request they address item2 that you raised.
Thank you once again for the input.
Regards

> Hello, sharkdevs,
> 
> I hope you had it easy to follow through.
> We have taken note and adjusted item 1 that you picked up.
> With item 2 we are looking on how to autogenerate a matching API.
> 
> Regards
> Steven/Wayodi
> 
> > * The body of your commit message should be separated from the header by 1 blank line. Ensure you change that.
> > * I cannot run the API from the Documentation link given. This is because the example values given do not match what the API Accepts. check the screenshot below
> >   ![image](https://user-images.githubusercontent.com/14871749/51174892-f5a16900-18c9-11e9-8af3-0ff50a29eab0.png)

Collins33 commented:
Great job on the task and the simplicity by which the instructions to set it up have been made. It works perfectly on my machine
